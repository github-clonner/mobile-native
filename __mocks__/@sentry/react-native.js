export const captureMessage = jest.fn();
export default {};
