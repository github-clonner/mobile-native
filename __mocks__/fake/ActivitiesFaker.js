
function load(count, container) {

  let activities = [...Array(count)].map((_, i) => {
    const code = 'activityguid' + i;
    return {
      attachment_guid: false,
      blurb: false,
      container_guid:code,
      custom_data:false,
      custom_type:false,
      rowKey: 'something' + i,
      description:"Congratulations! ",
      edited:"",
      guid:code,
      mature:false,
      time_created: "1522036284",
      ownerObj:{
        guid: "824853017709780997",
        type: "user",
        subtype: false,
        time_created: "1522036284",
        // getAvatarSource: () => {
        //   return {
        //     source:'http://thisisaurl'
        //   }
        // }
      },
      shouldBeBlured: jest.fn(),
      message:"Message",
      title:'TITLE',
      owner_guid:"824853017709780997",
      parent_guid:"838106762591510528",
      perma_url:false,
      thumbnail_src:false,
      type:"activity",
      wire_totals: {
        tokens: 1000000000000000000
      },
      containerObj: container
      // _list: {
      //   viewed: {
      //     viewed: new Map([["1019155171608096768",true]]),
      //     addViewed: () => {
      //       return;
      //     }
      //   }
      // },
      // getThumbSource: () => {
      //   return {
      //     source:'http://thisisaurl'
      //   }
      // }
    }
  });

  return {
    'activities': activities,
    'load-next': 'aaaaaa',
    'load-previous': 'aaaaaa'
  }

}

/**
 * comment service data faker
 * @param {integer} count
 */
export function activitiesServiceFaker() {
  return { load: load}
}

