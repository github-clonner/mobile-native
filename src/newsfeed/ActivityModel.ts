import { runInAction, action, observable, decorate } from 'mobx';
import FastImage from 'react-native-fast-image';
import { FlatList, Alert } from 'react-native';

import BaseModel from '../common/BaseModel';
import UserModel from '../channel/UserModel';
import wireService from '../wire/WireService';
import sessionService from '../common/services/session.service';
import {
  setPinPost,
  deleteItem,
  unfollow,
  follow,
  update,
} from '../newsfeed/NewsfeedService';
import api from '../common/services/api.service';

import { GOOGLE_PLAY_STORE, MINDS_CDN_URI, MINDS_URI } from '../config/Config';
import i18n from '../common/services/i18n.service';
import logService from '../common/services/log.service';
import entitiesService from '../common/services/entities.service';
import type { ThumbSize } from '../types/Common';
import type GroupModel from 'src/groups/GroupModel';

type Thumbs = Record<ThumbSize, string>;

/**
 * Activity model
 */
export default class ActivityModel extends BaseModel {
  // Observable properties
  @observable pinned: boolean = false;
  @observable time_created: string = '';
  @observable message: string = '';
  @observable title: string = '';
  @observable mature: boolean = false;
  @observable edited: '0' | '1' = '0';

  // decorated observables
  'is:following': boolean;
  'thumbs:down:count': number;
  'thumbs:up:count': number;
  'comments:count': number;
  'thumbs:down:user_guids': Array<number>;
  'thumbs:up:user_guids': Array<number>;

  containerObj?: GroupModel;
  remind_object?: ActivityModel;
  ownerObj!: UserModel;
  listRef?: FlatList<any>;
  thumbnails?: Thumbs;
  paywall: '1' | '' = '';
  paywall_unlocked: boolean = false;
  guid: string = '';
  subtype: string = '';
  entity_guid: string = '';
  owner_guid: string = '';
  custom_type?: string;
  custom_data?: Array<any> | any;
  nsfw?: Array<number>;
  flags?: any;
  reminds: number = 0;
  impressions: number = 0;
  perma_url?: string;
  cinemr_guid?: string;
  thumbnail_src?: string;
  dontPin?: boolean;
  boosted?: boolean;
  wire_threshold?: {
    type: 'token' | 'money';
    min: number;
  } | null;
  _preview?: boolean;
  /**
   * Set programatically to indicate that the parent is marked as mature
   */
  is_parent_mature?: boolean;

  /**
   * Mature visibility flag
   */
  @observable mature_visibility: boolean = false;

  /**
   * Is visible in flat list
   */
  @observable is_visible: boolean = true;

  /**
   *  List reference setter
   */
  set _list(value) {
    this.__list = value;

    // the reminded object need access to the metadata service too
    if (this.remind_object) {
      this.remind_object._list = value;
    }
  }

  /**
   *  List reference getter
   */
  get _list() {
    return this.__list;
  }

  toPlainObject() {
    const plainEntity = super.toPlainObject();
    if (plainEntity.remind_object && plainEntity.remind_object.__list) {
      delete plainEntity.remind_object.__list;
    }

    return plainEntity;
  }

  /**
   * Child models
   */
  childModels() {
    return {
      ownerObj: UserModel,
      remind_object: ActivityModel,
    };
  }

  /**
   * Get the activity thumb source
   * {uri: 'http...'}
   * @param {string} size
   */
  getThumbSource(size: ThumbSize = 'medium') {
    // for gif use always the same size to take adventage of the cache (they are not resized)
    if (this.isGif()) {
      size = 'medium';
    }

    if (this.thumbnails && this.thumbnails[size]) {
      return { uri: this.thumbnails[size], headers: api.buildHeaders() };
    }

    // fallback to old behavior
    if (this.paywall || this.paywall_unlocked) {
      return {
        uri: MINDS_URI + 'fs/v1/thumbnail/' + this.entity_guid,
        headers: api.buildHeaders(),
      };
    }
    if (this.custom_type === 'batch') {
      return {
        uri: MINDS_CDN_URI + 'fs/v1/thumbnail/' + this.entity_guid + '/' + size,
        headers: api.buildHeaders(),
      };
    }
    return {
      uri: MINDS_CDN_URI + 'fs/v1/thumbnail/' + this.guid + '/' + size,
      headers: api.buildHeaders(),
    };
  }

  isGif(): boolean {
    return this.custom_data && this.custom_data[0] && this.custom_data[0].gif;
  }

  /**
   * Preload thumb on image cache
   */
  preloadThumb(size: ThumbSize = 'medium') {
    FastImage.preload([this.getThumbSource(size)]);
  }

  shouldBeBlured(): boolean {
    const user = sessionService.getUser();

    if (user && user.mature) {
      return false;
    }

    if (this.nsfw !== undefined) {
      let res = [1, 2, 4].filter((nsfw) => this.nsfw!.indexOf(nsfw) > -1)
        .length;
      if (res) {
        return true;
      }
    }

    if (typeof this.flags !== 'undefined') {
      return !!this.flags.mature;
    }

    if (typeof this.mature !== 'undefined') {
      return !!this.mature;
    }

    if (
      typeof this.custom_data !== 'undefined' &&
      typeof this.custom_data[0] !== 'undefined'
    ) {
      return !!this.custom_data[0].mature;
    }

    if (this.custom_data !== undefined) {
      return !!this.custom_data.mature;
    }

    return false;
  }

  /**
   * Get activity text
   */
  get text() {
    return this.message || '';
  }

  @action
  toggleMatureVisibility() {
    if (GOOGLE_PLAY_STORE) {
      return;
    }
    this.mature_visibility = !this.mature_visibility;

    if (this.remind_object && this.remind_object.mature) {
      this.remind_object.mature_visibility = this.mature_visibility;
    }
  }

  @action
  setVisible(value: boolean) {
    this.is_visible = value;
  }

  /**
   * Increment the comments counter
   */
  @action
  incrementCommentsCounter() {
    this['comments:count']++;
  }

  /**
   * Decrement the comments counter
   */
  @action
  decrementCommentsCounter() {
    this['comments:count']--;
  }

  /**
   * Unlock the activity and update data on success
   */
  @action
  async unlock(ignoreError = false) {
    try {
      const result: object | false = await wireService.unlock(this.guid);
      if (result) {
        // all changes should be atomic (trigger render only once)
        runInAction(() => {
          // create a new model because we need the child models
          const model: ActivityModel = ActivityModel.create(result);
          Object.assign(this, model);
        });
      }
      return result;
    } catch (err) {
      if (!ignoreError) {
        Alert.alert(err.message);
      }
      return false;
    }
  }

  @action
  async togglePin() {
    // allow owners only
    if (!this.isOwner()) {
      return;
    }

    try {
      this.pinned = !this.pinned;
      await setPinPost(this.guid, this.pinned);
    } catch (e) {
      this.pinned = !this.pinned;
      Alert.alert(i18n.t('errorPinnedPost'));
    }
  }

  @action
  async deleteEntity() {
    try {
      await deleteItem(this.guid);
      this.removeFromList();
      entitiesService.deleteFromCache(this.urn);
    } catch (err) {
      logService.exception('[ActivityModel]', err);
      throw err;
    }
  }

  @action
  async toggleFollow() {
    const method = this['is:following'] ? unfollow : follow;
    try {
      await method(this.guid);
      runInAction(() => {
        this['is:following'] = !this['is:following'];
      });
    } catch (err) {
      logService.exception('[OffsetFeedListStore]', err);
      throw err;
    }
  }

  @action
  async updateActivity(data: any = {}) {
    const entity: any = this.toPlainObject();

    if (data) {
      for (const field in data) {
        entity[field] = data[field];
      }
    }

    // call update endpoint
    await update(entity);

    // update instance properties
    this.update(data);

    this.setEdited(entity.message);
  }

  @action
  setEdited(message: string) {
    this.message = message;
    this.edited = '1';
  }
}

/**
 * Define model observables
 */
decorate(ActivityModel, {
  //@ts-ignore
  'thumbs:down:count': observable,
  'thumbs:up:count': observable,
  'comments:count': observable,
  'thumbs:down:user_guids': observable,
  'thumbs:up:user_guids': observable,
});
