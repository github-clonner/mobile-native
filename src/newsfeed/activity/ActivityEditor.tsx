import React, { Component } from 'react';

import { View, TextInput, StyleSheet, Alert } from 'react-native';

import Button from '../../common/components/Button';

import { CommonStyle } from '../../styles/Common';
import HashtagService from '../../common/services/hashtag.service';
import autobind from '../../common/helpers/autobind';
import Colors from '../../styles/Colors';
import logService from '../../common/services/log.service';
import testID from '../../common/helpers/testID';
import CapturePosterFlags from '../../capture/CapturePosterFlags';
import i18n from '../../common/services/i18n.service';
import ThemedStyles from '../../styles/ThemedStyles';
import ActivityModel from '../ActivityModel';

type PropsType = {
  entity: ActivityModel;
  toggleEdit: Function;
};

type StateType = {
  text: string;
  saving: boolean;
  wire_threshold: any;
  nsfw: Array<number>;
  time_created: number;
};

export default class ActivityEditor extends Component<PropsType, StateType> {
  state: StateType = {
    text: '',
    saving: false,
    nsfw: [],
    wire_threshold: null,
    time_created: 0,
  };

  componentWillMount() {
    this.setState({
      text: this.props.entity.message,
      wire_threshold: this.props.entity.wire_threshold || null,
      nsfw: this.props.entity.nsfw || [],
      time_created: parseInt(this.props.entity.time_created, 10),
    });
  }

  update = async () => {
    if (
      HashtagService.slice(this.state.text).length > HashtagService.maxHashtags
    ) {
      //if hashtag count greater than 5
      Alert.alert(
        i18n.t('capture.maxHashtags', {
          maxHashtags: HashtagService.maxHashtags,
        }),
      );
      return false;
    }

    const data: any = {
      message: this.state.text,
      time_created: this.formatTimeCreated(),
    };

    data.nsfw = [...this.state.nsfw];

    data.wire_threshold = this.state.wire_threshold;
    if (data.wire_threshold) {
      data.paywall = data.wire_threshold.min > 0;
    } else {
      data.paywall = false;
    }

    try {
      this.setState({ saving: true });
      await this.props.entity.updateActivity(data);
      this.props.toggleEdit(false);
    } catch (err) {
      logService.exception('[ActivityEditor] update', err);
      Alert.alert(
        i18n.t('sorry'),
        i18n.t('errorMessage') + '\n' + i18n.t('activity.tryAgain'),
        [{ text: i18n.t('ok'), onPress: () => {} }],
        { cancelable: false },
      );
    } finally {
      this.setState({ saving: false });
    }
  };

  cancel = () => {
    this.props.toggleEdit(false);
  };

  @autobind
  onNsfwChange(nsfw: Array<number>) {
    this.setState({
      nsfw,
    });
  }

  @autobind
  onLocking(wire_threshold: any) {
    this.setState({ wire_threshold });
  }

  onScheduled = (timeCreated: number | string) => {
    this.setState({ time_created: parseInt(String(timeCreated), 10) });
  };

  formatTimeCreated = () => {
    let time_created: string | number = this.state.time_created;
    if (time_created) {
      time_created = new Date(this.state.time_created).getTime();
      time_created = Math.floor(time_created / 1000);
    }
    return time_created;
  };

  renderFlagsView() {
    return (
      <View style={[CommonStyle.rowJustifyStart, CommonStyle.paddingTop]}>
        <CapturePosterFlags
          //@ts-ignore
          hideShare={true}
          hideHash={true}
          lockValue={this.state.wire_threshold}
          nsfwValue={this.state.nsfw}
          onNsfw={this.onNsfwChange}
          onLocking={this.onLocking}
          timeCreatedValue={
            this.state.time_created
              ? new Date(this.state.time_created * 1000)
              : null
          }
          onScheduled={this.onScheduled}
        />
      </View>
    );
  }

  /**
   * Render
   */
  render() {
    const theme = ThemedStyles.style;

    return (
      <View style={[CommonStyle.flexContainer, CommonStyle.padding]}>
        <TextInput
          style={[
            theme.fullWidth,
            theme.backgroundTertiary,
            theme.padding,
            theme.borderRadius2x,
            theme.colorPrimaryText,
          ]}
          multiline={true}
          numberOfLines={4}
          onChangeText={(text) => this.setState({ text })}
          value={this.state.text}
          {...testID('Post editor input')}
        />
        <View style={styles.buttonBar}>
          {this.renderFlagsView()}

          <View style={[CommonStyle.rowJustifyEnd, CommonStyle.paddingTop]}>
            <Button
              text={i18n.t('cancel')}
              onPress={this.cancel}
              containerStyle={theme.paddingHorizontal2x}
              {...testID('Post editor cancel button')}
            />
            <Button
              text={i18n.t('save')}
              containerStyle={[theme.marginLeft2x, theme.paddingHorizontal2x]}
              onPress={this.update}
              disabled={this.state.saving}
              {...testID('Post editor save button')}
            />
          </View>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  buttonBar: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  nsfw: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },
  nsfwLabel: {
    color: Colors.explicit,
    fontWeight: '700',
  },
});
