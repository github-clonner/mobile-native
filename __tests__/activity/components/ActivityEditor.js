import 'react-native';
import React from 'react';

import { shallow } from 'enzyme';

import ActivityEditor from '../../../src/newsfeed/activity/ActivityEditor';
import { activitiesServiceFaker } from '../../../__mocks__/fake/ActivitiesFaker';
import ActivityModel from '../../../src/newsfeed/ActivityModel';
import apiService from '../../../src/common/services/api.service';

jest.mock('../../../src/common/services/api.service');

describe('Activity editor component', () => {
  let user, comments, entity, screen, newsfeed, toggleEdit, activity;
  beforeEach(() => {
    const navigation = { navigate: jest.fn() };
    let activityResponse = activitiesServiceFaker().load(1);
    activity = ActivityModel.create(activityResponse.activities[0]);
    toggleEdit = jest.fn();
    screen = shallow(
      <ActivityEditor
        entity={activity}
        toggleEdit={toggleEdit}
        navigation={navigation}
        newsfeed={newsfeed}
      />,
    );

    //jest.runAllTimers();
  });

  it('renders correctly', async () => {
    screen.update();
    expect(screen).toMatchSnapshot();
  });

  it('should have a Text input', async () => {
    screen.update();

    expect(screen.find('TextInput')).toHaveLength(1);
  });

  it('should set the default value when init, call toggle and submit', async () => {
    screen.update();

    apiService.post.mockResolvedValue({ activity: {} });

    let instance = screen.instance();
    expect(instance.state.text).toBe('Message');
    const spy = jest.spyOn(activity, 'updateActivity');
    screen.find('Button').at(1).props().onPress();

    expect(spy).toHaveBeenCalled();
  });

  it('should set the default value when init, call toggle and cancel', async () => {
    screen.update();
    let instance = screen.instance();
    expect(instance.state.text).toBe('Message');
    const spy = jest.spyOn(instance.props, 'toggleEdit');
    screen.find('Button').at(0).props().onPress();
    expect(spy).toHaveBeenCalled();
  });

  it('should set the text', async () => {
    screen.update();
    let instance = screen.instance();
    expect(instance.state.text).toBe('Message');

    screen.find('TextInput').forEach((child) => {
      child.simulate('changeText', 'data');
    });

    const spy = jest.spyOn(instance.props, 'toggleEdit');
    screen.find('Button').at(0).props().onPress();
    expect(spy).toHaveBeenCalled();
    expect(instance.state.text).toBe('data');
  });
});
