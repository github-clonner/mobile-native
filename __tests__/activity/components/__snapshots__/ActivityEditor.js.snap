// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity editor component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "padding": 5,
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    multiline={true}
    numberOfLines={4}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
    testID="Post editor input"
    underlineColorAndroid="transparent"
    value="Message"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "flex-start",
          },
          Object {
            "paddingTop": 5,
          },
        ]
      }
    >
      <inject-with-capture(CapturePosterFlags)
        hideHash={true}
        hideShare={true}
        lockValue={null}
        nsfwValue={Array []}
        onLocking={[Function]}
        onNsfw={[Function]}
        onScheduled={[Function]}
        timeCreatedValue={2018-03-26T03:51:24.000Z}
      />
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "flex-end",
          },
          Object {
            "paddingTop": 5,
          },
        ]
      }
    >
      <Button
        loading={false}
        onPress={[Function]}
        testID="Post editor cancel button"
        text="Cancel"
      />
      <Button
        containerStyle={
          Array [
            undefined,
            undefined,
          ]
        }
        disabled={false}
        loading={false}
        onPress={[Function]}
        testID="Post editor save button"
        text="Save"
      />
    </View>
  </View>
</View>
`;
